### PR TITLE
feat(storage): add the domain service apis for recording storage intent from the cli

### PR DESCRIPTION
**Pull request by wallyworld**
The CLI can be used to attach / detach storage from a unit, add storage to a unit, or import storage into a model. This PR adds the domain service layer methods to support those operations. The state layer is not implemented yet; this will be done in a follow up PR. Also, storage destroy is not included because that will be done in the removal domain later.

The service methods are added to 2 domains - application and storage. The application domain gets the methods used to make storage changes involving a unit. The storage domain gets the method to import storage.

As is being done for other entities, storage Name and ID are given their own types.

The state methods for attaching, detaching storage are declared to return a large number of domain specific errors which the mongo implementation doesn't have. These numerous errors cover things from the storage not being alive to the count limits being violated etc. I question slightly the need to do this given the service layer doesn't use them and they are just passed up to the facade caller, which will need to wrap them. There's no logic decisions based on having specific error types so the same outcome would be achieved by loosening the api contract to just return generic errors.

## QA steps

Just unit tests for now.

## Links

**Jira card:** [JUJU-7467](https://warthogs.atlassian.net/browse/JUJU-7467)

### CODE REVIEW

File: core/unit/id.go [hunk 1]
```diff
@@ -23,6 +23,16 @@ func NewUUID() (UUID, error) {
...
 
+// MustNewUUID is a convince function for generating a new unit uuid
+// and panics on error. Only used in tests.
+func MustNewUUID() UUID {
+	uuid, err := NewUUID()
+	if err != nil {
+		panic(err)
+	}
+	return uuid
+}
```
Comment (SimonRichardson) [inline core/unit/id.go:34]: This should be in `core/unit/testing` see the existing pattern for charms and applications.

File: domain/application/service/storage.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+// The following error types can be expected:
+// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+// - [applicationerrors.StorageAttachmentAlreadyExistsError]: when the attachment already exists.
+// - [applicationerrors.WrongStorageOwnerError]: when the storage owner does not match the unit.
+// - [applicationerrors.UnitNotAlive]: when the unit is not alive.
+// - [applicationerrors.StorageNotAlive]: when the storage is not alive.
+// - [applicationerrors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [applicationerrors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [applicationerrors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AttachStorage(ctx context.Context, storageID storage.ID, unitName unit.Name) error {
```
Comment (SimonRichardson) [inline domain/application/service/storage.go:63]: We should validate the ID and the unit name, before attempting a request on state.

File: domain/application/service/storage.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+// The following error types can be expected:
+// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+// - [applicationerrors.UnitNotAlive]: when the unit is not alive.
+// - [applicationerrors.StorageNotAlive]: when the storage is not alive.
+// - [applicationerrors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [applicationerrors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [applicationerrors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AddStorageForUnit(
+	ctx context.Context, storageName storage.Name, unitName unit.Name, stor storage.Directive,
+) ([]storage.ID, error) {
```
Comment (SimonRichardson) [inline domain/application/service/storage.go:82]: Validate here.

File: domain/application/service/storage.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		return nil, errors.Capture(err)
+	}
+	return s.st.AddStorageForUnit(ctx, storageName, unitUUID, stor)
+}
+
+// DetachStorageForUnit detaches the specified storage from the specified unit.
+// The following error types can be expected:
+// - [applicationerrors.UnitNotFound]: when the unit does not exist.
+// - [applicationerrors.StorageNotDetachable]: when the type of storage is not detachable.
+func (s *Service) DetachStorageForUnit(ctx context.Context, storageID storage.ID, unitName unit.Name) error {
```
Comment (SimonRichardson) [inline domain/application/service/storage.go:94]: Validate input.

File: domain/storage/service/service.go [hunk 5]
```diff
@@ -38,7 +45,7 @@ func NewService(st State, logger logger.Logger, registryGetter storage.ModelStor
 // Deprecated: This method will be removed once the storage registry is fully
 // implemented in each service.
 func (s *Service) GetStorageRegistry(ctx context.Context) (internalstorage.ProviderRegistry, error) {
-	registry, err := s.registryGetter.GetStorageRegistry(ctx)
+	registry, err := s.StorageService.registryGetter.GetStorageRegistry(ctx)
```
Comment (SimonRichardson) [inline domain/storage/service/service.go:48]: This feels like a smell to me.

File: domain/storage/service/storage.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	filesystemInfo := storage.FilesystemInfo{Pool: arg.Pool}
+
+	// If the storage provider supports filesystems, import the filesystem,
+	// otherwise import a volume which will back a filesystem.
+	invalidatorFunc, err := credentialInvalidatorGetter()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	callCtx := envcontext.WithCredentialInvalidator(ctx, invalidatorFunc)
```
Comment (SimonRichardson) [inline domain/storage/service/storage.go:90]: This makes me very sad.

File: domain/storage/service/storage.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+	var attr map[string]any
+	if len(poolDetails.Attrs) > 0 {
+		attr = transform.Map(poolDetails.Attrs, func(k, v string) (string, any) { return k, v })
+	}
```
Comment (SimonRichardson) [inline domain/storage/service/storage.go:95]: Just remove the if, it'll be a no-op in the end.

File: domain/storage/service/storage.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		filesystemInfo.BackingVolume = &internalstorage.VolumeInfo{
+			HardwareId: info.HardwareId,
+			WWN:        info.WWN,
+			Size:       info.Size,
+			VolumeId:   info.VolumeId,
+			Persistent: info.Persistent,
+		}
+		filesystemInfo.Size = info.Size
+	}
```
Comment (SimonRichardson) [inline domain/storage/service/storage.go:142]: This should be a couple of methods. It's doing way too much.

File: domain/storage/service/storage.go (hunk 6 above)
Comment (SimonRichardson) [reply domain/storage/service/storage.go:90]: Also, I think if the credential is invalid, we should poison the provider and the registry.

File: domain/storage/types.go [hunk 9]
```diff
@@ -72,3 +72,16 @@ func DefaultStoragePools(registry storage.ProviderRegistry) ([]*storage.Config,
 	}
 	return result, nil
 }
+
+// ModelDetails describes details about a model.
+type ModelDetails struct {
+	ModelUUID      string `db:"uuid"`
+	ControllerUUID string `db:"controller_uuid"`
+}
```
Comment (SimonRichardson) [inline domain/storage/types.go:80]: Remove db tags (copy and paste from above).

File: internal/storage/name.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Validate returns an error if the ID is invalid.
+// The returned error is an InvalidStorageID error.
+func (id ID) Validate() error {
+	if !validStorageID.MatchString(id.String()) {
+		return errors.Errorf("%w: %q", InvalidStorageID, id)
+	}
+	return nil
+}
```
Comment (SimonRichardson) [inline internal/storage/name.go:79]: Can this not go in core/storage or does it collide?

File: domain/storage/service/storage.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+// ImportFilesystem associates a filesystem (either native or volume backed) hosted by a cloud provider
+// with a new storage instance (and storage pool) in a model.
+// The following error types can be expected:
+// - [coreerrors.NotSupported]: when the importing the kind of storage is not supported by the provider.
+// - [storageerrors.InvalidPoolNameError]: when the supplied pool name is invalid.
+func (s *StorageService) ImportFilesystem(
+	ctx context.Context, credentialInvalidatorGetter envcontext.ModelCredentialInvalidatorGetter, arg ImportStorageParams,
+) (internalstorage.ID, error) {
+	if arg.Kind != internalstorage.StorageKindFilesystem {
+		// TODO(axw) implement support for volumes.
```
Comment (manadart) [inline domain/storage/service/storage.go:48]: I think we've done this right?

File: domain/storage/service/storage.go (hunk 6 above)
Comment (wallyworld) [reply domain/storage/service/storage.go:90]: This code is literally copied from the existing facade implementation. This whole credential validation thing needs to be considered holistically, not just in this one call site. The current implementation effectively pauses the model workers until the bad credential is updated. Any thinking on implementation changes can happen outside this PR.

File: domain/storage/service/storage.go (hunk 8 above)
Comment (wallyworld) [reply domain/storage/service/storage.go:142]: The whole block of code was copied directly from the facade. I can clean up a bit.

File: internal/storage/name.go (hunk 10 above)
Comment (wallyworld) [reply internal/storage/name.go:79]: Unlike other things, `storage` lives under `internal` since some things need a provider context. In 3.6, there is no `core/storage` at all; it is top level. 
For 4.0, `juju/juju/storage` is moved to `internal/storage`.

In 4.0, we added one thing to `core/storage`. We could add this to core as well. The split brain becomes a bit annoying but the stuff that needs provider artefacts needs to stay where it is.

File: domain/storage/service/storage.go (hunk 11 above)
Comment (wallyworld) [reply domain/storage/service/storage.go:48]: The CLI command is `import-filesystem`. We do support importing volume backed filesystems. There's never been though fully implemented and tested support for importing block device storage. Hence the todo.

File: domain/storage/service/service.go (hunk 5 above)
Comment (wallyworld) [reply domain/storage/service/service.go:48]: The whole method is going away as soon as the facade cut over is done. 
It's marked as deprecated for that reason. But for now we need it to serve the facade.

File: core/unit/id.go (hunk 1 above)
Comment (wallyworld) [reply core/unit/id.go:34]: yeah, totally forgot that

File: domain/storage/service/storage.go (hunk 7 above)
Comment (wallyworld) [reply domain/storage/service/storage.go:98]: `transform.Map` returns an empty map when the input is nil which could be considered "bad" behaviour
YMMV. I wish `transform.Map` didn't do that.

in mongo land, Attrs was stored as `map[string]any` so no transformation was needed and tests in particular could assume Attrs was nil if the input config had nil Attrs. Without the check all the tests would need to transform the struct under test to have an empty map.

File: domain/storage/types.go (hunk 9 above)
Comment (wallyworld) [reply domain/storage/types.go:80]: ugh, cut and paste fail

File: internal/storage/name.go (hunk 10 above)
Comment (wallyworld) [reply internal/storage/name.go]: Moved the Name and ID to core and left the current stuff in internal/storage where it is.

But there's a bunch of stuff like StorageKind etc that should also be moved to core.

File: domain/storage/service/storage.go (hunk 6 above)
Comment (wallyworld) [reply domain/storage/service/storage.go]: In other places we have punted to using a no op context via `envcontext.WithoutCredentialInvalidator()`. ie we have just kicked the can down the road. So I'll do the same here to be consistent and we can holistically address the issue across all services which need this.

File: domain/storage/service/params.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"github.com/juju/juju/core/storage"
+	internalstorage "github.com/juju/juju/internal/storage"
+)
+
+// ImportStorageParams contains the parameters for importing storage into a model.
+type ImportStorageParams struct {
+	// Kind is the kind of the storage to import.
+	Kind internalstorage.StorageKind
```
Comment (tlm) [inline domain/storage/service/params.go:14]: Does it seem correct that we have this type/enum defined in an internal package?

File: domain/storage/service/storage.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	}
+	return filesystemInfo, nil
+}
+
+func (s *StorageService) importFilesystemFromProvider(ctx envcontext.ProviderCallContext, provider internalstorage.Provider, cfg *internalstorage.Config, providerID string, resourceTags map[string]string) (*storage.FilesystemInfo, error) {
+	filesystemSource, err := provider.FilesystemSource(cfg)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	filesystemImporter, ok := filesystemSource.(internalstorage.FilesystemImporter)
```
Comment (tlm) [inline domain/storage/service/storage.go:122]: There is a todo that exists on this interface to clean up this casting smell. Can we do it now and just add a func to the provider or source?

File: domain/storage/service/storage.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+func (s *StorageService) importVolumeFromProvider(
+	ctx envcontext.ProviderCallContext, provider internalstorage.Provider, cfg *internalstorage.Config,
+	providerID string, resourceTags map[string]string,
+) (*storage.FilesystemInfo, error) {
+	volumeSource, err := provider.VolumeSource(cfg)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	volumeImporter, ok := volumeSource.(internalstorage.VolumeImporter)
```
Comment (tlm) [inline domain/storage/service/storage.go:151]: Same casting smell as above.

File: domain/storage/service/storage_test.go [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+	return ctrl
+}
+
+func (s *storageSuite) service(c *gc.C) *Service {
+	return NewService(s.state, loggertesting.WrapCheckLog(c), modelStorageRegistryGetter(func() storage.ProviderRegistry {
+		return s.registry
+	}))
+}
+
+func (s *storageSuite) TestImportFilesystemValidate(c *gc.C) {
```
Comment (tlm) [inline domain/storage/service/storage_test.go:73]: Ideally these could all just be `c.Check` and this could also be wrapped up into tests slice if you  wanted.

File: domain/storage/service/storage_test.go [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+	cfg, err := storage.NewConfig("elastic", "elastic", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	s.provider.EXPECT().FilesystemSource(cfg).Return(filesystemImporter{
+		MockFilesystemSource:   s.filesystemSource,
+		MockFilesystemImporter: s.filesystemImporter,
+	}, nil)
+
+	s.state.EXPECT().GetStoragePoolByName(gomock.Any(), "elastic").Return(domainstorage.StoragePoolDetails{}, storageerrors.PoolNotFoundError)
+	s.state.EXPECT().GetModelDetails().Return(domainstorage.ModelDetails{
+		ModelUUID:      coretesting.ModelTag.Id(),
```
Comment (tlm) [inline domain/storage/service/storage_test.go:114]: I don't think we should be bringing tag's down here. Can we just gen these once in setup and save them on the suite ?

File: domain/storage/service/storage_test.go [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+		ControllerUUID: coretesting.ControllerTag.Id(),
+	}, nil)
+	s.filesystemImporter.EXPECT().ImportFilesystem(gomock.Any(), "provider-id", map[string]string{
+		"juju-model-uuid":      coretesting.ModelTag.Id(),
+		"juju-controller-uuid": coretesting.ControllerTag.Id(),
+	}).Return(storage.FilesystemInfo{
+		FilesystemId: "provider-id",
```
Comment (tlm) [inline domain/storage/service/storage_test.go:121]: Would it be adventurous to use a different id here in the tests from that of the provider so when a failure happens it is possible to tell which value is which in test output?

File: domain/storage/state/storagetypes.go [hunk 18]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+// These structs represent the persistent storage schema in the database.
+
+type ModelDetails struct {
```
Comment (tlm) [inline domain/storage/state/storagetypes.go:8]: Do these types need to be exported and are we supposed to be prefixing with db or is that just some packages?

**Review by tlm — COMMENTED** [review]
Thanks @wallyworld there has been a fair amount of work put into this.

File: core/storage/name.go [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+	validStorageID   = regexp.MustCompile("^(" + StorageNameSnippet + ")/" + NumberSnippet + "$")
+	validStorageName = regexp.MustCompile(StorageNameSnippet)
+)
+
+// Name represents a storage name.
+type Name string
+
+// NewName returns a new Name. If the name is invalid, an InvalidStorageName error
+// will be returned.
+func NewName(name string) (Name, error) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:38]: This follows what we have already done for unit name. I have no opinion either way, just want to be consistent and the unit name was done first.
```
// NewName returns a new Name. If the name is invalid, an InvalidUnitName error
// will be returned.
func NewName(name string) (Name, error) {
	n := Name(name)
	return n, n.Validate()
}
```

File: domain/storage/service/storage_test.go (hunk 16 above)
Comment (wallyworld) [reply domain/storage/service/storage_test.go:114]: This is the universal pattern we've been using in every test package to date. Same with controller UUID. The pattern goes back years. If we want to change it, we could add something to the core testing package alongside the model and controller tags. Generating and saving attrs on the test suite works but is unnecessary overhead.

File: domain/storage/service/storage_test.go (hunk 17 above)
Comment (wallyworld) [reply domain/storage/service/storage_test.go:121]: sure

File: domain/storage/state/storagetypes.go (hunk 18 above)
Comment (wallyworld) [reply domain/storage/state/storagetypes.go:8]: Yeah, this was added here prematurely - deleting in this PR. Should have been in the next PR. Doesn't need to be exported. Some packages use db prefix. Many don't. We are inconsistent. The types already in storage package don't.

File: core/storage/name.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+const (
+	// InvalidStorageName represents an invalid storage name.
+	InvalidStorageName = errors.ConstError("invalid storage name")
+	// InvalidStorageID represents an invalid storage id.
+	InvalidStorageID = errors.ConstError("invalid storage id")
+)
+
+var (
+	validStorageID   = regexp.MustCompile("^(" + StorageNameSnippet + ")/" + NumberSnippet + "$")
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:29]: We use the current pattern in over 100 places in the code base. If we want to change the approach, can we have a conversation about doing it holistically.

File: core/storage/name.go [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+	return n, n.Validate()
+}
+
+// String returns the Name as a string.
+func (n Name) String() string {
+	return string(n)
+}
+
+// Validate returns an error if the Name is invalid.
+// The returned error is an InvalidStorageName error.
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:49]: yup

File: core/storage/name.go [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	return nil
+}
+
+// ID represents a storage ID which is a name with a sequence number.
+type ID string
+
+// NewID returns a new ID. If the id is invalid, an InvalidStorageID error
+// will be returned.
+func NewID(id string) (ID, error) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:62]: Same reply :-)
`NewID` is already used in more places than `NewName`
If we want to change it, let's do it holistically.

File: core/storage/name.go [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+// String returns the Name as a string.
+func (n Name) String() string {
+	return string(n)
+}
+
+// Validate returns an error if the Name is invalid.
+// The returned error is an InvalidStorageName error.
+func (n Name) Validate() error {
+	if !validStorageName.MatchString(n.String()) {
+		return errors.Errorf("%w: %q", InvalidStorageName, n)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:52]: sure

File: domain/storage/service/params.go (hunk 12 above)
Comment (wallyworld) [reply domain/storage/service/params.go:14]: It never used to be internal and wasn't supported to be but it got moved by someone. 
We now need to go in and move a bunch of stuff out from internal storage and add it to core. This is one of the things.

File: domain/storage/service/storage_test.go (hunk 15 above)
Comment (wallyworld) [reply domain/storage/service/storage_test.go:73]: There were only 3 of them so the overhead of a slice didn't seem worth it. Made them a Check

File: domain/storage/service/storage_test.go (hunk 16 above)
Comment (wallyworld) [reply domain/storage/service/storage_test.go]: For these tests, I've used generated uuids instead of the testing ones. We can look to holistically adopt a consistent approach over time.

File: core/storage/name.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (tlm) [reply core/storage/name.go:38]: The problem is it gives a potential reader a false sense of security about usability that can't be enforced.

You can construct a name by either `NewName("test")` or `Name("test")`.  But when you use the resultant value n no guarantees can ever be provided that the value in n is valid until you call 'Validate'. Only the initial constructor of the value knows if the value is valid.

New implies this the correct mechanism for constructing the type and no other mechanism should be used. By doing this we force valid constructors of the type to always validate when they may also know the value is valid. i.e state reading. We haven't adopted this pattern in lots of other places so I would say the one example pointed to is against the grain not with it.

Constructing a new type from virgin materials with validation is a parsing/unmarshalling step not a `New` construction.

File: core/storage/name.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (tlm) [reply core/storage/name.go:62]: Same comment as above. This is new code and it is against the convention we have established. Look at UUID values add to core.

File: core/storage/name.go [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+	if !validStorageName.MatchString(n.String()) {
+		return errors.Errorf("validating storage name %q", n).Add(InvalidStorageName)
+	}
+	return nil
+}
+
+// ID represents a storage ID which is a name with a sequence number.
+type ID string
+
+// NewID returns a new ID. If the id is invalid, an InvalidStorageID error
```
Comment (tlm) [inline core/storage/name.go:60]: nit: `[InvalidStorageID]'

File: domain/application/service/storage.go [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+// - [github.com/juju/juju/core/storage.InvalidStorageName]: when the storage name is not valid.
+// - [github.com/juju/juju/domain/storage/errors.StorageNotFound] when the storage doesn't exist.
+// - [github.com/juju/juju/domain/application/errors.UnitNotFound]: when the unit does not exist.
+// - [github.com/juju/juju/domain/application/errors.UnitNotAlive]: when the unit is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNotAlive]: when the storage is not alive.
+// - [github.com/juju/juju/domain/application/errors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
+func (s *Service) AddStorageForUnit(
+	ctx context.Context, storageName corestorage.Name, unitName coreunit.Name, stor storage.Directive,
```
Comment (tlm) [inline domain/application/service/storage.go:114]: is `stor` a typo?

File: domain/application/service/storage_test.go [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+	service *Service
+}
+
+var _ = gc.Suite(&storageSuite{})
+
+func (s *storageSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.mockState = NewMockState(ctrl)
+
+	s.service = &Service{
```
Comment (tlm) [inline domain/application/service/storage_test.go:37]: Should we not have a New func for this ?

File: domain/application/state/storage.go [hunk 27]
```diff
@@ -73,3 +77,55 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
 	}
 	return nil
 }
+
+// GetStorageUUIDByID returns the UUID for the specified storage, returning an error
```
Comment (tlm) [inline domain/application/state/storage.go:81]: In an ideal world  this would be introduced in a seperate PR to keep this ones size down and concerns isolated.

File: domain/application/state/types.go [hunk 28]
```diff
@@ -876,3 +877,8 @@ func decodeWorkloadStatus(s int) (application.WorkloadStatusType, error) {
 		return -1, errors.Errorf("unknown status %d", s)
 	}
 }
+
+type minimalStorage struct {
```
Comment (tlm) [inline domain/application/state/types.go:881]: minimalStorage to me doesn't tell me what  this thing is or does. Just a nit but it sounds like a Juju concept on first read that I didn't know about.

Comment (tlm) [reply domain/application/state/types.go:881]: What about storageIdMapping ?

File: domain/storage/service/params.go (hunk 12 above)
Comment (tlm) [reply domain/storage/service/params.go:14]: Ta would be nice to get this done in another PR

File: domain/storage/service/storage.go (hunk 13 above)
Comment (tlm) [reply domain/storage/service/storage.go:122]: fair. The func comment was to wrap this up like we do with environproviders and the provider tracker where you get a getter that you can request an interface off of. 

Removes all of this if then that from the structure and make the func very procedural. Just a nice to have.

**Review by tlm — APPROVED** [review]
Giving this conditional approval. It is very good work bringing this over. Most of my comments are style related more then anything else.

File: domain/storage/service/storage.go (hunk 13 above)
Comment (wallyworld) [reply domain/storage/service/storage.go:122]: Trouble is the interface here is returned by calling a getter on the provider, and the capability of the result is dependent on the type of storage we are using. It will be a lot of work if we want to look at changing the implementation. Given this PR is primarily concerned with moving existing logic down from the facade layer to the service layer, let's keep it at that and look at provider implementation concerns separately.

File: core/storage/name.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:38]: Sure. We will have to go back and redo the existing NewName and NewID implementations used elsewhere.

File: domain/application/service/storage.go (hunk 25 above)
Comment (wallyworld) [reply domain/application/service/storage.go:114]: no, i just didn't want to use "storage" and "store" implies something else. 
Renamed to "directive".

File: domain/application/state/types.go (hunk 28 above)
Comment (wallyworld) [reply domain/application/state/types.go:881]: We use the "minimal" naming patter elsewhere as well.
In the next PR, life and other attributes will be added, but not all of them, just a "minimal" set.
I'll rename this one to "storageDetails" and we can see what people like best and adopt a consistent approach.

File: domain/application/service/storage_test.go (hunk 26 above)
Comment (wallyworld) [reply domain/application/service/storage_test.go:37]: sure

File: core/storage/name.go [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+	// storage names (without the storage instance sequence number).
+	storageNameSnippet = "(?:[a-z][a-z0-9]*(?:-[a-z0-9]*[a-z][a-z0-9]*)*)"
+	// numberSnippet is a non-compiled regexp that can be composed with other
+	// snippets for validating small number sequences.
+	numberSnippet = "(?:0|[1-9][0-9]*)"
+)
+
+const (
+	// InvalidStorageName represents an invalid storage name.
+	InvalidStorageName = errors.ConstError("invalid storage name")
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/storage/name.go:23]: In core, we haven't tended to use separate errors packages. What's here also matches the convention adopted for the `InvalidUnitName` error in core unit.

**Review by ycliuhw — APPROVED** [review]
LGTM, ty!
